### PR TITLE
Unified logger for Azure ML, App Insights, and Console

### LINKED ISSUE #17: AML Observability Library
Requirements:

1. The model scoring application should be able to log metrics and trace events to App Insights.
2. Will also be to nice to have the same code to log metrics to both App Insights and Azure ML during training.

### LINKED ISSUE #29: Console Logger approach
It would maybe more straight forward to just add a StreamHandler (https://docs.python.org/3/howto/logging.html#configuring-logging) to the logger used in the Observability Singleton to activate Log to console. 

That may be also true for AzureMLLogger as well, the only custom logic we need there is the metric's logging mechanism to the AML run.

Currently referring to [this branch](https://github.com/liupeirong/MLOpsManufacturing/tree/h2floh/17_logger) until #23 is merged

### PR DESCRIPTION
**Pull request by liupeirong**
This library uses a single interface to log messages and metrics to Azure ML, App Insights, and Console. It also supports dependency tracing in App Insights. 

- [x] Text and metric logging for all 3 loggers are implemented.
- [x] Unit tests passed for all loggers.
- [x] Dependency tracing/spans are supported in App Insights.
- [x] Published to Test PyPI (https://test.pypi.org/project/azureml-appinsights-logger/). 
- [x] Run both standalone and in Azure ML experiment.
- [x] CI/CD pipeline to automate linting, unit testing, and packaging.

### CODE REVIEW

File: common/azureml_appinsights_logger/azureml_appinsights_logger/azureml_logger.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        :param description: An optional metric description.
+        :type description: str
+        """
+        if name != "":
+            self.run.log(
+                name, value, description
+            ) if log_parent is False or self.run.parent is None \
+                else self.run.parent.log(name, value, description)
+
+    def log(self, description="", severity=Severity.INFO):
```
Comment (mokarian) [inline common/azureml_appinsights_logger/azureml_appinsights_logger/azureml_logger.py:35]: As we discussed, since we are already passing the massages in ConsoleLogger, this method should just be 
```python
def log(self, description="", severity=Severity.INFO):
      pass
```

File: common/azureml_appinsights_logger/azureml_appinsights_logger/logger_interface.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+class Severity:
+    DEBUG = 10
+    INFO = 20
+    WARNING = 30
+    ERROR = 40
+    CRITICAL = 50
+
+
+class LoggerInterface(Tracer):
```
Comment (mokarian) [inline common/azureml_appinsights_logger/azureml_appinsights_logger/logger_interface.py:14]: Why are we inheriting Tracer in this Interface?

File: common/azureml_appinsights_logger/azureml_appinsights_logger/logger_interface.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    def log_metric(self, name, value, description, log_parent):
+        pass
+
+    def log(self, name, value, description, severity, log_parent):
+        pass
+
+    def exception(self, exception):
+        pass
+
+    def span(self, name='span'):
```
Comment (mokarian) [inline common/azureml_appinsights_logger/azureml_appinsights_logger/logger_interface.py:25]: I believe the following methods should not be in  this interface:

- span
- start_span
- end_span
- current_span
- add_attribute_to_current_span
- list_collected_spans

These methods are only used in AppInsightsLogger. It should not be implemented by all the Loggers. hence they should not be living here.

File: common/azureml_appinsights_logger/azureml_appinsights_logger/observability.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        this is a reference that can be used in case
+        any of the built in functions of the loggers is required
+        :param logger_class:
+        :return: a logger class
+        """
+        for logger in self._loggers.loggers:
+            if type(logger) is type(logger_class):
+                return logger
+
+    def span(self, name='span'):
```
Comment (mokarian) [inline common/azureml_appinsights_logger/azureml_appinsights_logger/observability.py:115]: As we discussed, the following methods should not be here. 
These should only be the concern of **AppInsightsLogger**.
- span
- start_span
- end_span
- current_span
- add_attribute_to_current_span
- list_collected_spans

Since only AppInsightsLogger support these methods, it's misleading and against our abstraction to be in Observability Class. 

One way to cal this method with Observability is  :
```python
observability.get_logger(AppInsightsLogger("")).start_span(..)
...
observability.get_logger(AppInsightsLogger("")).span(..)
```

File: samples/image-classification-tensorflow/devops_pipelines/01-pull-request.yml [hunk 5]
```diff
@@ -18,6 +18,8 @@ pr:
     include:
     - main
   paths:
+    include:
```
Comment (mokarian) [inline samples/image-classification-tensorflow/devops_pipelines/01-pull-request.yml]: Is this change related to this project?

File: common/azureml_appinsights_logger/README.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  - opencensus-ext-httplib==0.7.*
+```
+ 
+## 2. Configure environment variables to control logger behavior
+
+```bash
+# AppInsights logger is enabled if this is a valid connection string
+APPLICATIONINSIGHTS_CONNECTION_STRING = ''
+# Console logger is enabled if set to 'true'
+LOG_TO_CONSOLE = 'false'
```
Comment (mokarian) [inline common/azureml_appinsights_logger/README.md:31]: Should this be default to true?

File: common/azureml_appinsights_logger/README.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+# DEBUG, INFO, WARNING, ERROR, CRITICAL, default is WARNING
+LOG_LEVEL = 'WARNING' 
+# Probability 0.0 -> 1.0
+LOG_SAMPLING_RATE = '1.0'
+# Probability 0.0 -> 1.0
+TRACE_SAMPLING_RATE = '1.0'
+# Frequency in seconds to send metric to App Insights, default is 15
+METRICS_EXPORT_INTERVAL = '15'
+# Whether to log App Insights standard machine metrics, CPU, memory etc.
+ENABLE_STANDARD_METRICS = 'true'
```
Comment (mokarian) [inline common/azureml_appinsights_logger/README.md:42]: Should the default value be false?
Are we going to capture the metrics from the machine that is running our script?
I found this polluting our logs in Application insights.

**Review by mokarian — CHANGES_REQUESTED** [review]
Looks good. I am so glad to see this is growing by your team. Happy to collaborate more.
Added few comments that we discussed.

File: common/azureml_appinsights_logger/azureml_appinsights_logger/azureml_logger.py (hunk 1 above)
Comment (liupeirong) [reply common/azureml_appinsights_logger/azureml_appinsights_logger/azureml_logger.py:35]: This is resolved by deriving a flag "log_text_to_aml" from the inversion of "log_to_console" in env_variables.py.

File: common/azureml_appinsights_logger/azureml_appinsights_logger/observability.py (hunk 4 above)
Comment (liupeirong) [reply common/azureml_appinsights_logger/azureml_appinsights_logger/observability.py:115]: Thought about this one. I'd prefer to keep the current way because:

- Other loggers could implement span in the future. 
- If multiple loggers implement span, and the caller wants to log to multiple destinations, then they'd need to get_logger and call the logging methods on each one of them, defying the purpose of having a unified logger.

File: common/azureml_appinsights_logger/azureml_appinsights_logger/logger_interface.py (hunk 3 above)
Comment (liupeirong) [reply common/azureml_appinsights_logger/azureml_appinsights_logger/logger_interface.py:25]: See below comment.

File: common/azureml_appinsights_logger/azureml_appinsights_logger/logger_interface.py (hunk 2 above)
Comment (liupeirong) [reply common/azureml_appinsights_logger/azureml_appinsights_logger/logger_interface.py:14]: The logger that implements tracing, App Insights logger for now, inherits the parent tracer object.

File: samples/image-classification-tensorflow/devops_pipelines/01-pull-request.yml (hunk 5 above)
Comment (liupeirong) [reply samples/image-classification-tensorflow/devops_pipelines/01-pull-request.yml]: moved this to a separate PR.

File: common/azureml_appinsights_logger/azureml_appinsights_logger/observability.py (hunk 4 above)
Comment (h2floh) [reply common/azureml_appinsights_logger/azureml_appinsights_logger/observability.py:115]: While I understand @mokarian concern from a coding perspective, I intentionally put the methods in the Interface that it becomes convenient to use the library as an AML developer. The "heavy load" should be abstracted, even if that results in not getting recorded the expected data.

Also the observability lib/package will automatically activate AppInsights and other loggers based on ENV values and dispatch/chain the calls to the different types of loggers. So on the AML and ML script side you don't need to know about what loggers are or will be active.

File: common/azureml_appinsights_logger/README.md (hunk 6 above)
Comment (h2floh) [reply common/azureml_appinsights_logger/README.md:31]: Hi @mokarian,

as I went deeper into the [python logging mechanism](https://docs.python.org/3/howto/logging.html#handlers) we are not using it very clean right now (opportunity for a refactoring/PR :) ). 

Actually we should just add a console/stream handler for console logging. 

The AppInsights logger works as a handler. So we [add the handler](https://github.com/liupeirong/MLOpsManufacturing/blob/0bd6873209cd4ff67d14318116bc196ccef39a30/common/azureml_appinsights_logger/azureml_appinsights_logger/appinsights_logger.py#L51) to the python logging and then calling the logging methods like [here](https://github.com/liupeirong/MLOpsManufacturing/blob/0bd6873209cd4ff67d14318116bc196ccef39a30/common/azureml_appinsights_logger/azureml_appinsights_logger/appinsights_logger.py#L102).

Even if several handlers are active you only need to call the `logger.<criticality>(msg)` once to get it recorded to all handlers. What happened when I did this with the console logger was due to our current structure [here](https://github.com/liupeirong/MLOpsManufacturing/blob/0bd6873209cd4ff67d14318116bc196ccef39a30/common/azureml_appinsights_logger/azureml_appinsights_logger/observability.py#L90) that `logger.<criticality>(msg)` was called twice  (once in the console logger object and then again in the app insights logger object) which resulted in doubled log entries on the console and on AppInsights, because 2 handlers were active at each point in time.

Therefore I just modified the console logger into almost the same thing as the AML logger, which just does a `print` to STDOUT. The only logger class which is currently using python `logger.<criticality>(msg)` is the App Insights class and contains only the App Insights handler, so we don't get duplicated messages into AppInsights.

However in the current situation, if you then run the scripts on AML service and the `LOG_TO_CONSOLE` is true, you get doubled STDOUT entries in the AML studio logger view. That's why I set default to `false` here.

The idea was that we wanted to get the library to life first and then take care about these details. There is room for improvement as you can see. I would think that we need following:

- rehaul the [Observability Class](https://github.com/liupeirong/MLOpsManufacturing/blob/0bd6873209cd4ff67d14318116bc196ccef39a30/common/azureml_appinsights_logger/azureml_appinsights_logger/observability.py#L51) to add handlers to the python logging mechanism as needed
- Calls to `logger.<criticality>(msg)` only happen once in Observability Class, rest is handled by the handlers
- AML Logger only needs to provide the metrics logging to AML service, the SDTOUT logging for AML Studio Log view can be captured by the Console Handler (I think it's called stream handler)

CC @liupeirong

File: common/azureml_appinsights_logger/README.md (hunk 7 above)
Comment (h2floh) [reply common/azureml_appinsights_logger/README.md:42]: Sure wouldn't mind if this is set to `false`. I found it useful looking and traversing through the Application Map that these insights are already populated by default.

**Review by mokarian — APPROVED** [review]

**Review by h2floh — APPROVED** [review]
